### PR TITLE
last overhaul of ZJSON

### LINKED ISSUE #1780: update zjson to handle type values
For the time being, the app can display type values as ZSON strings, but at some point will want the type values encoded using zjson type encodings.

### LINKED ISSUE #2185: zjson type values need to be updated
zjson type values should be represented as a well-defined native json structure instead of strings.  This way, clients do not need to parse the string to create the type-value structure.

### LINKED ISSUE #2376: zjson type values need to be updated
zjson type values are currently encoded as strings.  This means that the client would have to parse the string to recreate a complex type structure.  Instead, we should update zjson to use nested JSON structure to encode the Z type just like it does for complex values allowing native JSON-oritned clients (like zealot and the python library) to more easily decode a Z type value.

### LINKED ISSUE #2525: zjson should use ZSON primitive values
zjson currently uses the TZNG format for primitive values.  We should change this to ZSON but it has impact on zealot and the python client so we will stage this appropriately.

### PR DESCRIPTION
**Pull request by mccanne**
This commit implements one last overhaul to the ZJSON spec
to bring it into alignment with where the Zed design landed.
Importantly, types and type values are consistently encoded
using JSON structure so that clients can have an easier time
parsing and managing the Zed type context for a stream of
ZJSON records.  This also means that type-values are encoded
with JSON structure instead of using ZSON type strings that
in turn needed to be parsed.

This change follows the compiler AST pattern for Zed types
but keeps the simple encoding of values based on nested
JSON arrays of strings.  The types are now more verbose but
easier to parse at the client.  Also, the aliases hack was
removed in favor of Zed's model of embedded typedefs.

We also eliminated the use of package joe and removed it
as this was the last remaining vestige of joe-ness.

Closes #1780 
Closes #2185 
Closes #2376 
Closes #2525

### CODE REVIEW

**Comment by mccanne** [comment]
I put this up as a draft since we need to update the python client (hence failing tests) and the zealot lib in the app.   Otherwise (aside from enum), ZJSON is officially done!

File: docs/formats/zjson.md [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+```
+{
+  "kind": "enum",
+  "symbols": [ <string>, <string>, ... ]
+}
+```
+
+#### Type Definition
+
+A type definitions is encoded as a binding between a name and a Zed type
```
Comment (jameskerr) [inline docs/formats/zjson.md:224]:
```suggestion
A type definition is encoded as a binding between a name and a Zed type
```

File: docs/formats/zjson.md [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+  "name": <id>,
+  "type": <type>,
+}
+```
+where `<id>` is a JSON string representing the newly defined type name
+and `<type>` is a recursively encoded type.
+If `<id>` is a non-integer string, then it is a "type name" that is a user visible
+type name.  If it is an integer, then it is is not user visible and is used  
+exclusively to collate firt-class Zed type values in the value array with
+their corresponding type.
```
Comment (jameskerr) [inline docs/formats/zjson.md:238]:
```suggestion
If `<id>` is a non-integer string, then it is a "type name" that is a user-visible
type name.  If it is an integer, then it is not user-visible and is used  
exclusively to collate first-class Zed type values in the values array with
their corresponding type.
```

File: docs/formats/zjson.md [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+
+#### Type Name
+
+A type reference is encoded as refer to a previously defined type definition.
```
Comment (jameskerr) [inline docs/formats/zjson.md:242]:
```suggestion
A type reference is encoded to refer to a previously defined type definition 
```

**Review by jameskerr — COMMENTED** [review]
Thanks Steve, this looks like it will make the code similar to reason about. I like it. Here are few typo/wording suggestions. I'll get a client updated soon.

File: docs/formats/zjson.md [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+```
+where `<id>` is a JSON string representing a previously defined type name.
+
+### Value Encoding
+
+The primitive values comprising an arbitrarily complex Zed data value are encoded
+as a JSON array of strings mixed with nested JSON arrays whose structure
+conforms to the nested structure of the value's schema as follows:
+* each record, array, and set is encoded as a JSON array of its composite values,
+* a union is encoded as a string of the form '<selector:<value>>' where `selector`
```
Comment (philrz) [inline docs/formats/zjson.md:258]:
```suggestion
* a union is encoded as a string of the form `<selector:<value>>` where `selector`
```

File: docs/formats/zjson.md [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+[ "hello, world", ["1","2","3","4"], ["1:foo", "0:10" ] ]
+```
+
+## Framing ZJSON objects
+
+A sequence of ZJSON objects may be framed in two primary ways.
+
+First, they can simply be a sequence of newline delimited JSON where
+each object is transmitted as a single line terminated with a newline character,
+e.g., the [zq](https://github.com/brimdata/zed) CLI command writes its
```
Comment (philrz) [inline docs/formats/zjson.md:279]:
```suggestion
e.g., the [zq](https://github.com/brimdata/zed/tree/main/cmd/zq) CLI command writes its
```

File: docs/formats/zjson.md [hunk 6]
```diff
@@ -0,0 +1,492 @@
+# Zed over JSON
+
+* [ZJSON](#zjson)
+  + [Type Encoding](#type-encoding)
+    - [Record Type](#record-type)
+    - [Array Type](#array-type)
+    - [Set Type](#set-type)
+    - [Union type](#union-type)
+    - [Enum Type](#enum-type)
+  + [Alias Encoding](#alias-encoding)
```
Comment (philrz) [inline docs/formats/zjson.md:10]:
```suggestion
    - [Enum Type](#enum-type)
    - [Type Definition](#type-definition)
    - [Type Name](#type-name)
```

**Comment by jameskerr** [comment]
Maybe make the Map type return an array of arrays in the values? Right now it's flat:


```
    {
      kind: 'map',
      key_type: { kind: 'primitive', name: 'string' },
      val_type: { kind: 'primitive', name: 'int64' }
    } [ 'best', '1', 'worst', '0' ]
```

File: ppl/zqd/ztests/python.yaml [hunk 7]
```diff
@@ -14,14 +15,62 @@ script: |
...
+        i8:0,
+        i16:0,
+        i32:0,
+        i64:0,
+        dur:0m,
+        tim:1970-01-01T00:00:00Z,
+        f64:0.,
+        boo:false,
+        // Uncomment next line after https://github.com/brimdata/zed/issues/2533 is fixed.
+        // byt:0x00,
```
Comment (nwt) [inline ppl/zqd/ztests/python.yaml:59]:
```suggestion
        byt:0x00,
```

File: docs/formats/zjson.md [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+  + [Value Encoding](#value-encoding)
+* [Framing ZJSON objects](#framing-zjson-objects)
+* [Example](#example)
+
+The Zed data model is based on richly typed records and a deterministic column order,
```
Comment (nwt) [inline docs/formats/zjson.md:15]:
```suggestion
The Zed data model is based on richly typed records with a deterministic column order,
```

File: docs/formats/zjson.md [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+As a comparison, Python's `json` module handles the 64-bit integer to full
+precision, but loses precision on the floating point timestamp.
+Also, as mentioned, it is at the whim of a JSON implementation whether
+or not the order of object keys is preserved.
+
+While JSON is well suited for data exchange of generic information, it is not
+so appropriate for a structured data model like Zed.
+That said, JSON can be used as an encoding format for Zed by mapping Zed data
+onto a JSON-based protocol.  This allows clients like web apps or
+electron apps to receive and understand Zed and, with the help of client
```
Comment (nwt) [inline docs/formats/zjson.md:71]:
```suggestion
Electron apps to receive and understand Zed and, with the help of client
```

File: docs/formats/zjson.md [hunk 10]
```diff
@@ -0,0 +1,492 @@
...
+libraries like [zealot](https://github.com/brimdata/brim/tree/master/zealot),
+to be enabled with rich, structured Zed types that are implemented on top of
```
Comment (nwt) [inline docs/formats/zjson.md:73]:
```suggestion
to manipulate the rich, structured Zed types that are implemented on top of
```

File: docs/formats/zjson.md [hunk 11]
```diff
@@ -0,0 +1,492 @@
...
+recursively with small-integer type identifiers.
+
+Since Zed steams are self describing and type information is embedded
+in the stream itself, the embedded types are likewise encoded in the
+ZJSON format.
+
+A ZJSON stream is defined as a sequence of JSON objects where each object
+represents a Zed value.  Each object includes an identifier that denotes
+its type, or _schema_.  A schema generically refers to the type of the
+Zed record that is defined by a given JSON object.
```
Comment (nwt) [inline docs/formats/zjson.md:100]: This paragraph refers to "a Zed value" and then, later, "the Zed record". I think it will read better sticking to one or the other.

File: docs/formats/zjson.md [hunk 12]
```diff
@@ -0,0 +1,492 @@
...
+{
+  "schema": <id>,
+  "values": [ <val>, ... [ <val>, ... ] ... ]
+  "types": [ <type>, <type>, ... ]
+}
+```
+
+### Type Encoding
+
+The type format follows the terminology in the ZSON spec, where primitive types
```
Comment (nwt) [inline docs/formats/zjson.md:127]:
```suggestion
The type format follows the terminology in the [ZSON spec](zson.md), where primitive types
```

File: docs/formats/zjson.md [hunk 13]
```diff
@@ -0,0 +1,492 @@
...
+#### Enum Type
+
+A Zed enum type is a JSON object of the form
+```
+{
+  "kind": "enum",
+  "symbols": [ <string>, <string>, ... ]
+}
+```
+
```
Comment (nwt) [inline docs/formats/zjson.md:221]: Don't forget maps!
~~~
#### Map Type

A Zed map type is defined by a JSON object of the form
```
{
  "kind": "map",
  "key_type": <type>,
  "val_type": <type>
}
```

~~~

File: docs/formats/zjson.md [hunk 14]
```diff
@@ -0,0 +1,492 @@
...
+#### Type Definition
+
+A type definitions is encoded as a binding between a name and a Zed type
+and represents a new type so named.  A type definition type has the form:
```
Comment (nwt) [inline docs/formats/zjson.md:225]: For consistency.
```suggestion
and represents a new type so named.  A type definition type has the form
```

File: docs/formats/zjson.md [hunk 15]
```diff
@@ -0,0 +1,492 @@
...
+
+The primitive values comprising an arbitrarily complex Zed data value are encoded
+as a JSON array of strings mixed with nested JSON arrays whose structure
+conforms to the nested structure of the value's schema as follows:
+* each record, array, and set is encoded as a JSON array of its composite values,
+* a union is encoded as a string of the form '<selector:<value>>' where `selector`
+is an integer string representing the positional index in the union's list of
+types that specifies the type of `<value>`, which is a JSON string or array
+as described recursively herein,
+* a type value is encoded as its integer type identifier represented as a string, and
```
Comment (nwt) [inline docs/formats/zjson.md:262]:
```suggestion
as described recursively herein,
* a map is encoded as ...
* a type value is encoded as its integer type identifier represented as a string, and
```

File: docs/formats/zjson.md [hunk 16]
```diff
@@ -0,0 +1,492 @@
...
+and an array of union of string, and float64 --- might have a value that looks like this:
+```
+[ "hello, world", ["1","2","3","4"], ["1:foo", "0:10" ] ]
+```
+
+## Framing ZJSON objects
+
+A sequence of ZJSON objects may be framed in two primary ways.
+
+First, they can simply be a sequence of newline delimited JSON where
```
Comment (nwt) [inline docs/formats/zjson.md:277]:
```suggestion
First, they can simply be [newline delimited JSON (NDJSON)](http://ndjson.org/), where
```

File: docs/formats/zjson.md [hunk 17]
```diff
@@ -0,0 +1,492 @@
...
+each object is transmitted as a single line terminated with a newline character,
+e.g., the [zq](https://github.com/brimdata/zed) CLI command writes its
+ZJSON output as lines of NDJSON.
+
+Second, the objects may be encoded in a JSON array embedded in some other
+JSON-framed protocol, e.g., embedded in the the search results messages
+of the [zqd REST API](../../api/api.go).
+
+It is up to an implementation to determine how the ZJSON
```
Comment (nwt) [inline docs/formats/zjson.md:286]:
```suggestion
It is up to an implementation to determine how ZJSON
```

File: docs/formats/zjson.md [hunk 18]
```diff
@@ -0,0 +1,492 @@
...
+
+```
+{s:"hello",r:{a:1 (int32),b:2 (int32)} (=0)} (=1)
+{s:"world",r:{a:3,b:4}} (1)
+{s:"hello",r:{a:[1 (int32),2 (int32),3 (int32)] (=2)} (=3)} (=4)
+{s:"goodnight",r:{x:{u:"foo" (5=((string,int32)))} (=6)} (=7)} (=8)
+{s:"gracie",r:{x:{u:12 (int32)}}} (8)
+```
+
+This data is represented in ZJSON as follows;
```
Comment (nwt) [inline docs/formats/zjson.md:302]:
```suggestion
This data is represented in ZJSON as follows:
```

File: docs/formats/zjson.md [hunk 19]
```diff
@@ -0,0 +1,492 @@
...
+and `<type>` is a recursively encoded type.
+If `<id>` is a non-integer string, then it is a "type name" that is a user visible
+type name.  If it is an integer, then it is is not user visible and is used  
+exclusively to collate firt-class Zed type values in the value array with
+their corresponding type.
+
+#### Type Name
+
+A type reference is encoded as refer to a previously defined type definition.
+and has the form:
```
Comment (nwt) [inline docs/formats/zjson.md:243]: For consistency.
```suggestion
and has the form
```

File: ppl/zqd/worker.go [hunk 20]
```diff
@@ -38,7 +40,7 @@ func handleWorkerRootSearch(c *Core, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 	w.Header().Set("Content-Type", out.ContentType())
-	if err := srch.Run(r.Context(), store, out, req.MaxWorkers, c.conf.Worker); err != nil {
+	if err := srch.Run(r.Context(), store, out, req.MaxWorkers, c.conf.Worker, zctx); err != nil {
```
Comment (nwt) [inline ppl/zqd/worker.go:43]: Revert the changes to this file?

File: zio/zjsonio/ast.go [hunk 21]
```diff
@@ -0,0 +1,20 @@
+package zjsonio
+
+import (
+	"github.com/brimdata/zed/compiler/ast"
+	"github.com/brimdata/zed/pkg/unpack"
+)
+
+var unpacker = make(unpack.Reflector)
```
Comment (nwt) [inline zio/zjsonio/ast.go:8]:
```suggestion
var unpacker unpack.Reflector
```

File: zio/zjsonio/stream.go [hunk 22]
```diff
@@ -2,44 +2,133 @@ package zjsonio
...
+			return zng.SkipContainer
+		}
+		if typ == zng.TypeType {
+			typ, err := s.zctx.FromTypeBytes(bytes)
+			if err != nil {
+				// this shouldn't happen
+				return nil
+			}
+			_, t := s.typeID(typ)
+			if t != nil {
```
Comment (nwt) [inline zio/zjsonio/stream.go:127]:
```suggestion
			if _, t := s.typeID(typ); t != nil {
```

File: zio/zjsonio/types.go [hunk 23]
```diff
@@ -2,169 +2,159 @@ package zjsonio
...
-func encodeType(typ zng.Type) (joe.String, joe.Interface) {
 	switch typ := typ.(type) {
+	case *zng.TypeAlias:
+		name := typ.Name
+		t := e.encodeType(zctx, typ.Type)
+		e[typ] = name
+		return &ast.TypeDef{
+			Kind: "typedef",
+			Name: name,
+			Type: t,
```
Comment (nwt) [inline zio/zjsonio/types.go:29]:
```suggestion
		e[typ] = name
		return &ast.TypeDef{
			Kind: "typedef",
			Name: name,
			Type: e.encodeType(zctx, typ.Type),
```

File: zio/zjsonio/reader.go [hunk 24]
```diff
@@ -46,50 +45,37 @@ func (r *Reader) Read() (*zng.Record, error) {
...
 			return nil, err
 		}
-		r.mapper[v.Id] = recType
+	}
+	typ, ok := r.decoder[rec.Schema]
+	if !ok {
+		return nil, fmt.Errorf("undefined schema ID: %s", rec.Schema)
+	}
+	if !zng.IsRecordType(typ) {
+		return nil, fmt.Errorf("zjson outer type is not a record: %s", zson.FormatType(typ))
```
Comment (nwt) [inline zio/zjsonio/reader.go:62]: s/zjson/ZJSON/ in all the error strings?

**Comment by mccanne** [comment]
> What do you think about making the Map type return an array of arrays in the values? Right now it's flat:
> 
> ```
>     {
>       kind: 'map',
>       key_type: { kind: 'primitive', name: 'string' },
>       val_type: { kind: 'primitive', name: 'int64' }
>     } 
> 
>     [ 'best', '1', 'worst', '0' ]
> ```

So, you want   `[ ['best', '1'],['worst', '0'] ]`, right?

File: docs/formats/zjson.md (hunk 11 above)
Comment (mccanne) [reply docs/formats/zjson.md:100]: Good catch.  ZJSON actually works for the generic Zed model (any stream of values not just records) so I adopted "Zed value" over record.  (Not that we've implemented this... :-)

File: pkg/unpack/reflector.go [hunk 25]
```diff
@@ -13,19 +13,31 @@ var zero reflect.Value
...
 func (r Reflector) Add(template interface{}) Reflector {
+	return r.addAs(template, "")
+}
+
+// Override the unpack value tag with the as argument.
+func (r Reflector) AddAs(template interface{}, as string) Reflector {
+	return r.addAs(template, as)
+}
+
+func (r Reflector) addAs(template interface{}, as string) Reflector {
```
[the remark this answers is not in the document]
Comment (mccanne) [reply pkg/unpack/reflector.go:32]: doh!

File: zio/zjsonio/types.go (hunk 23 above)
Comment (mccanne) [reply zio/zjsonio/types.go:29]: This might not be clear in the spec but if the type name changes the later visitor in the depth-first traversal should win so I think it's better to leave it as it is..

File: zio/zjsonio/ast.go (hunk 21 above)
Comment (nwt) [reply zio/zjsonio/ast.go]: This was a bad suggestion because `unpack.Reflector` is a map.

New suggestion: Use `unpack.New`!

Comment (nwt) [reply zio/zjsonio/ast.go:8]: Newer suggestion: Go back to your original code, which I've done in 4d3ff279f46b9b72a64312cb800ef19a694d07b2.

File: docs/formats/zjson.md [hunk 26]
```diff
@@ -0,0 +1,509 @@
...
+```
+where `<id>` is a JSON string representing a previously defined type name.
+
+### Value Encoding
+
+The primitive values comprising an arbitrarily complex Zed data value are encoded
+as a JSON array of strings mixed with nested JSON arrays whose structure
+conforms to the nested structure of the value's schema as follows:
+* each record, array, and set is encoded as a JSON array of its composite values,
+* a union is encoded as a string of the form `<selector:<value>>` where `selector`
```
Comment (nwt) [inline docs/formats/zjson.md:270]:
```suggestion
* a union is encoded as a string of the form `<selector>:<value>>` where `selector`
```

File: docs/formats/zjson.md [hunk 27]
```diff
@@ -0,0 +1,509 @@
...
+is an integer string representing the positional index in the union's list of
+types that specifies the type of `<value>`, which is a JSON string or array
+as described recursively herein,
+* a map is encoded as a list of keys of values where each the key and the value
+alternate in sequence as elements of an array that are recursively encoded,
```
Comment (nwt) [inline docs/formats/zjson.md:275]: Should be something like this.
```suggestion
* a map is encoded as a JSON array of two-element arrays of the form
`[ <key>, <value> ]` where `key` and `value` are recursively encoded,
```

File: zio/anyio/ztests/zjson-gz.yaml [hunk 28]
```diff
@@ -1,15 +0,0 @@
-zed: '*'
```
Comment (nwt) [inline zio/anyio/ztests/zjson-gz.yaml:1]: 920b67a1754c6eec237e6592ea7361d200e4c79f removed this and `zjson-nonewline.yaml`. Was that intentional?

**Review by nwt — APPROVED** [review]

File: zio/anyio/ztests/zjson-gz.yaml (hunk 28 above)
Comment (mccanne) [reply zio/anyio/ztests/zjson-gz.yaml:1]: They are the same as other tests and were hard to keep updated modulo auto-decompression etc that is already extensively tested.
